### PR TITLE
add BOM detection, language guessing, and multi-byte char counts

### PR DESCRIPTION
**Pull request by scholarsmate**
This adds Byte Order Mark (BOM) detection and language guessing to the file metrics header.  This also added character count data to the data profiler.

Screenshot 2023-09-28 at 14 45 41

Above we see Arabic (ar in ISO-639-1) detected and no BOM.

Screenshot 2023-09-28 at 14 45 58

Above we see Greek (el in ISO-639-1) detected and the UTF-16LE BOM.

Screenshot 2023-09-28 at 14 47 58

Here we see the ISO-639-1 code expanded to the detected langauge.

Screenshot 2023-09-28 at 14 46 26

The last set of figures are related to multi-byte character counts.

### CODE REVIEW

**Comment by stricklandrbls** [comment]
Looks like we need to start deviating from the layout of the file metrics. If the _Content Type_ is too long then it's a little difficult to read the BOM text.

bom-readability

**Comment by scholarsmate** [comment]
@stricklandrbls, I thought about that.  We can just take the last bit say `foo/bar` and just show `bar`, but have a tooltip that shows `foo/bar`.

**Comment by stricklandrbls** [comment]
Here's a more pronounced overflow example
image

**Comment by stricklandrbls** [comment]
@scholarsmate 

Also, I'm not sure what I need to edit within this text file I created using Greek characters. The BOM is accurately detected in the data editor but the language is appearing as `unknown`.

<details><summary><b>Screenshot:</b> <i>My text input command</i></summary>

bom-1

</details>

---

Then I actually used the data editor to input the BOM magic number for `0xFF 0xFE` for UTF-16-LE
<details><summary><b>Screenshot:</b> <code>hexdump -C</code> <i>output of file</i></summary>

bom-2

</details>

---

The data editor detects the BOM correctly but the language is 'unknown'.
<details><summary><b>Screenshot:</b> <i>FileMetrics Panel Info</i></summary>

bom-3

</details>

---

There is another file I created, again with Greek characters, and the BOM detection is correct but the language is detected as Korean:
<details><summary><b>Screenshot:</b> <code>/tmp/utf16-le.txt</code></summary>

bom-greek-korean

</details>
<details><summary><b>Screenshot:</b> <code>vim /tmp/greek-utf16le.txt</code> <i>showing Korean</i></summary>

vim-greek-korean

</details>

**Comment by stricklandrbls** [comment]
### BOM & Editor Encoding Discrepancy
When the BOM detection occurs, if it matches the available UTF encodings in the data editor (UTF-8 and UTF-16LE) then the editor's _Edit Encoding_ setting should be changed to match. Currently it remains on Latin-1.

### BOM Description Mismatches w/ Profiler
If the user edits the BOM bytes of a file, then opens the profiler to re-detect the BOM, those changes are not reflecting the in _File Metrics_ panel. This persists even after the profiler display has closed.

<details><summary><h4>Screenshot</h4></summary>

bom-profiler-panel-mismatch

</details>

**Comment by scholarsmate** [comment]
@stricklandrbls, the BOM in the file metics is computed at session creation time.  The BOM in the profiler is computed on demand and will therefore always reflect current changes.  We can force a BOM sync at profile time to get agreement, but I think this is a case where the juice isn't worth the squeeze.  Instead, I just won't show it in the FileMetrics header at all, but we can use the BOM that we get at session creation time to set the default encoding.  That will be a nice touch.  What do you think?

**Comment by scholarsmate** [comment]
@stricklandrbls, proposed changes made.

**Comment by stricklandrbls** [comment]
I tested the BOM and language detection functionality against the test files that were located within the Omega Edit test files. The data editor displayed the correct BOM and language so there must have been a binary data issue with the testfiles I created earlier.

I'm still seeing some weird behavior with the data editor and omega edit server and currently testing to see if it's local to this branch only.

File: src/dataEditor/dataEditorClient.ts [hunk 1]
```diff
@@ -443,13 +460,27 @@ export class DataEditorClient implements vscode.Disposable {
...
+            length
+          )
           await this.panel.webview.postMessage({
             command: MessageCommand.profile,
             data: {
               startOffset: startOffset,
               length: length,
               byteProfile: byteProfile,
               numAscii: numAscii(byteProfile),
+              characterCount: {
```
Comment (stricklandrbls) [inline src/dataEditor/dataEditorClient.ts:475]: I feel like the `characterCount` field should be a single object given that there's already a `CharacterCountData` class defined in the DataMetrics.svelte file. It will just have to be moved to a file in which both the extension and UI can import from.

File: src/svelte/src/components/DataMetrics/DataMetrics.svelte [hunk 2]
```diff
@@ -257,6 +268,16 @@ limitations under the License.
         case MessageCommand.profile:
           numAscii = msg.data.data.numAscii as number
           byteProfile = msg.data.data.byteProfile as number[]
+
+          // character count data
```
Comment (stricklandrbls) [inline src/svelte/src/components/DataMetrics/DataMetrics.svelte:272]: I wonder if the MessageCommand ticket that @nlewis05 is working may effect the data extraction from the vscode websocket messages and in turn effect this portion as well.

Or to clean this up, the `CharacterCountData` class can have an extraction function that takes the message as a argument

File: src/svelte/src/components/DataMetrics/DataMetrics.svelte [hunk 3]
```diff
@@ -491,6 +512,16 @@ limitations under the License.
...
+    <label for="char-count-bom">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;BOM: <span id="char-count-bom" class="nowrap">{characterCountData.byteOrderMark}</span></label>
+    <label for="char-count-bom-bytes">&nbsp;&nbsp;BOM Bytes: <span id="char-count-bom-bytes" class="nowrap">{characterCountData.byteOrderMarkBytes}</span></label>
+    <label for="char-count-single">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Single: <span id="char-count-single" class="nowrap">{characterCountData.singleByteCount}</span></label>
+    <label for="char-count-double">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Double: <span id="char-count-double" class="nowrap">{characterCountData.doubleByteCount}</span></label>
+    <label for="char-count-triple">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Triple: <span id="char-count-triple" class="nowrap">{characterCountData.tripleByteCount}</span></label>
+    <label for="char-count-quad">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Quad: <span id="char-count-quad" class="nowrap">{characterCountData.quadByteCount}</span></label>
+    <label for="char-count-invalid">&nbsp;&nbsp;&nbsp;&nbsp;Invalid: <span id="char-count-invalid" class="nowrap">{characterCountData.invalidBytes}</span></label>
```
Comment (stricklandrbls) [inline src/svelte/src/components/DataMetrics/DataMetrics.svelte:524]: Would the fields for the `CharacterCountData` ever change in the future? 
I think a `{#each ...}` would make it more idiomatic to the actual object.

Comment (scholarsmate) [reply src/svelte/src/components/DataMetrics/DataMetrics.svelte:524]: I don't plan on changing the fields in `CharacterCountData` as I have complete coverage of unicode done.

**Review by stricklandrbls — APPROVED** [review]
+1

File: src/svelte/src/components/Header/fieldsets/FileMetrics.svelte [hunk 4]
```diff
@@ -71,6 +72,9 @@ limitations under the License.
           if ('type' in msg.data.data) {
             $fileMetrics.type = msg.data.data.type
           }
+          if ('language' in msg.data.data) {
+            $fileMetrics.language = msg.data.data.language
```
Comment (mbeckerle) [inline src/svelte/src/components/Header/fieldsets/FileMetrics.svelte:76]: Honestly, I'm looking, but I'm not finding any code that does any language guessing calls here. 

Can you point out when and where exactly this code takes some action which provides the guess at the language?

File: src/dataEditor/dataEditorClient.ts [hunk 5]
```diff
@@ -248,14 +256,30 @@ export class DataEditorClient implements vscode.Disposable {
...
+        : 'unknown'
+      assert(data.byteOrderMark.length > 0, 'byteOrderMark is not set')
 
-      this.fileSize = createSessionResponse.hasFileSize()
-        ? (createSessionResponse.getFileSize() as number)
-        : 0
+      data.language = createSessionResponse.hasLanguage()
+        ? (createSessionResponse.getLanguage() as string)
+        : 'unknown'
+      assert(data.language.length > 0, 'language is not set')
```
Comment (scholarsmate) [inline src/dataEditor/dataEditorClient.ts:277]: @mbeckerle, here is where the Data Editor is getting the detected language.  This happens at session creation time and is detected by the Ωedit™ server (see: https://github.com/ctc-oss/omega-edit/blob/main/server/scala/serv/src/main/scala/com/ctc/omega_edit/grpc/EditorService.scala#L124).  When the session is created in the extension, it sends a message to the WebView front-end with the details for the display.

**Review by mbeckerle — CHANGES_REQUESTED** [review]
Some renaming and improved doc strings/comments would be helpful. Those are the only changes I'm requesting. 

You can take the rest of this comment with a grain of salt, perhaps TL;DR. 

I am not a UI programmer. 

This code feels like mostly "glue logic" to me, and that's always something I like to see generated or minimized, not written by hand. It strikes me that this code consists of a structure definition of  byteOrderMark and other metadata, and that the only actual "code" in this is estabilishing the subscriptions on changes to that for the visual presentation. The rest could all be generated from the structure. 

Seems to me there is a pattern here for how to interact (controller) with the omega edit server, (model) which has a bunch of details associated with dealing with it being an asynchronous server. Added to that is how to interact with the VSCode presentation layer (view) and that all this should be ruthlessly uniform, ideally, generated from declarative specs of what the model contains, and what UI component subscribes to and creates changes of those model parts. 

What I cannot see in this PR, which is after all, a small part of the system, is any separation of those pieces that makes any sense to me. It all feels jumbled, and snarled with the details of GPB for talking to the omega edit server, and HTTP-style post for talking to ... the UI?
But that's perhaps because I'm not working with the larger code body - I'm looking at trees and bushes, so can't see the forest. 

Maybe I just need to add a model change that shows up visually on the display, to force myself to learn better how this all hangs together. (A suggestion on such a feature is welcome.)

This is btw, just my usual reaction whenever I see the detail required and intricacies of UI-related code, (Tho, omega edit server adds to the complexity in this case) and as I am a a back-end/algorithm developer, you can take my comments with a grain of salt. 

The promise of reactive UI "kit" was, I always hoped, to make this sort of thing far more declarative than it has been historically.

File: src/dataEditor/dataEditorClient.ts (hunk 5 above)
Comment (scholarsmate) [reply src/dataEditor/dataEditorClient.ts:277]: In this patch so far, you can do a segment via the data profiler.  The profiler will detect the BOM at the beginning of the file and do the character counts in the desired segment on demand.  It doesn't currently re-calculate the language or the file-type, however as these are being done at session creation time right now (ref: https://github.com/ctc-oss/omega-edit/blob/main/server/scala/serv/src/main/scala/com/ctc/omega_edit/grpc/EditorService.scala#L115).

Language detection on a segment ought to be fairly straightforward to add to the data profiler.  We can stop doing it at session creation and do it on demand with the profiler.

For content-type I think we just leave that as-is (done only at session creation time).  If the content-type changes (maybe because we opened an empty file and are creating a file from whole cloth), I can add a refresh widget to refresh the content type based on file updates.

In summary, I agree that we ought to be able to do language detection on a segment and discontinue doing it at session creation time, and for content-type, I think a refresh mechanism will be the best bet.

To support these changes, I'll need to update Ωedit™ and cut a new release for the server-side support.

File: src/dataEditor/dataEditorClient.ts [hunk 6]
```diff
@@ -237,6 +234,17 @@ export class DataEditorClient implements vscode.Disposable {
       'checkpointPath is not set'
     )
 
+    let data = {
+      byteOrderMark: '',
+      changeCount: 0,
+      computedFileSize: 0,
```
[the remark this answers is not in the document]
Comment (scholarsmate) [reply src/dataEditor/dataEditorClient.ts:240]: We use these things all over the place in the Data Editor.  Maybe we actually need a glossary. Much of it is defined [here](https://github.com/ctc-oss/omega-edit/wiki#metrics) but we'll need the glossary to live in the wiki on this project as the Ωedit™ stuff is just a subset of the variables used in the Data Editor.

ByteOrderMark:  Here it's either None, UTF-8, UTF-16LE, UTF-16BE, UTF-32LE, and UTF-32BE (enum defined here: https://github.com/ctc-oss/omega-edit/blob/main/core/src/include/omega_edit/fwd_defs.h#L76, turned from enum to string (what the server sends to the extension) here: https://github.com/ctc-oss/omega-edit/blob/main/core/src/lib/utility.c#L319).  We use it to set the default encoding setting in the Settings header, and we also use it on the FileMetrics header for display.

ChangeCount: The number of applied changes (edit transactions)

ComputedFileSize: The file size with respect to the applied changes (Can be different than the DiskFileSize)

Type here is the detected Content-Type as determined by Ωedit™ (which in turn delegates the content detection to Apache Tikka)

UndoCount: The count of Undos available for undoing

We can create another issue for publishing a glossary.

File: src/dataEditor/dataEditorClient.ts [hunk 7]
```diff
@@ -443,13 +460,27 @@ export class DataEditorClient implements vscode.Disposable {
...
+          )
           await this.panel.webview.postMessage({
             command: MessageCommand.profile,
             data: {
               startOffset: startOffset,
               length: length,
               byteProfile: byteProfile,
               numAscii: numAscii(byteProfile),
+              characterCount: {
+                byteOrderMark: characterCount.getByteOrderMark(),
```
[the remark this answers is not in the document]
Comment (scholarsmate) [reply src/dataEditor/dataEditorClient.ts:476]: The BOM is related to the character counts as it informs how the bytes will be decoded into characters.  Would `characterCountInfo` or `characterCountData` be better for the name?

File: src/dataEditor/dataEditorClient.ts [hunk 8]
```diff
@@ -610,16 +641,16 @@ export class DataEditorClient implements vscode.Disposable {
...
       this.fileToEdit = fileToSave
-      this.fileSize = await getComputedFileSize(this.omegaSessionId)
+      const fileSize = await getComputedFileSize(this.omegaSessionId)
       await this.panel.webview.postMessage({
         command: MessageCommand.fileInfo,
         data: {
-          computedFileSize: this.fileSize,
-          diskFileSize: this.fileSize,
-          fileName: this.fileToEdit,
+          computedFileSize: fileSize,
```
[the remark this answers is not in the document]
Comment (scholarsmate) [reply src/dataEditor/dataEditorClient.ts:648]: That's correct.  A glossary will help.

File: src/svelte/src/components/DataMetrics/DataMetrics.svelte [hunk 9]
```diff
@@ -491,6 +512,16 @@ limitations under the License.
         >{((numAscii / sum) * 100).toFixed(2)}</span
       >
     </label>
+    </div>
+    <hr />
+  <div class="char-count">
```
[the remark this answers is not in the document]
Comment (scholarsmate) [reply src/svelte/src/components/DataMetrics/DataMetrics.svelte:517]: What other metadata are thinking would be added?  I think these character count fields are _final_ at least as far as Unicode is concerned.

File: src/svelte/src/components/Header/fieldsets/FileMetrics.svelte [hunk 10]
```diff
@@ -71,6 +72,9 @@ limitations under the License.
           if ('type' in msg.data.data) {
             $fileMetrics.type = msg.data.data.type
```
[the remark this answers is not in the document]
Comment (scholarsmate) [reply src/svelte/src/components/Header/fieldsets/FileMetrics.svelte:73]: I agree, but this is a VS Code (legacy?) messaging thing.  Not much we can do about this one.

File: src/svelte/src/components/Header/fieldsets/Settings.svelte [hunk 11]
```diff
@@ -28,8 +28,22 @@ limitations under the License.
...
 
   $: $bytesPerRow = $displayRadix === RADIX_OPTIONS.Binary ? 8 : 16
+
+  window.addEventListener('message', (msg) => {
+    switch (msg.data.command) {
+      case MessageCommand.fileInfo:
+      {
+        if ('byteOrderMark' in msg.data.data) {
+            const { byteOrderMark } = msg.data.data
+            if (byteOrderMark === 'UTF-8') $editorEncoding = 'utf-8'
```
[the remark this answers is not in the document]
Comment (scholarsmate) [reply src/svelte/src/components/Header/fieldsets/Settings.svelte:41]: FE FF is for UTF-16BE, and EF BB BF is the UTF-8 BOM.   UTF-8 can typically be _assumed_, but it can also be _explicit_ with the UTF-8 BOM (though you're right that byte order isn't even a thing in UTF-8 as it is in UTF-16 and UTF-32).

File: src/svelte/src/components/Header/fieldsets/Settings.svelte [hunk 12]
```diff
@@ -28,8 +28,22 @@ limitations under the License.
   } from '../../../stores'
   import FlexContainer from '../../layouts/FlexContainer.svelte'
   import { UIThemeCSSClass } from '../../../utilities/colorScheme'
+  import {MessageCommand} from "src/utilities/message";
 
   $: $bytesPerRow = $displayRadix === RADIX_OPTIONS.Binary ? 8 : 16
+
+  window.addEventListener('message', (msg) => {
```
[the remark this answers is not in the document]
Comment (scholarsmate) [reply src/svelte/src/components/Header/fieldsets/Settings.svelte:35]: Ωedit™ provides the model, the dataEditorClient is the controller, and all the Svelte stuff is the view.  The dataEditorClient is doing quite a lot of taking data from the Ωedit™ client and populating VS Code messages for sending to the WebView.  It's also taking the messages from the WebView and translating actions into Ωedit™ client calls.

We may be able to just take the ProtocolBuffer reply messages straight from the client and convert them (JSON-ify) in the controller and ship them to the view and perhaps we ought to in many cases, but I think that is outside the scope of this PR.  Perhaps we create another issue for looking into slimming down the controller by just marshaling messages between the model and the view and vice-versa.

**Comment by scholarsmate** [comment]
@mbeckerle, your overall criticism of the controller portion is fair.  Neither @stricklandrbls or I are "UI developers/designers", but we are computer scientists and are willing to tackle any layer in the stack and we're confident that whatever we develop will work.  We've read _Design Patterns_ we've read Scott Meyers, Herb Sutter, Martin Fowler, and the like.  We know about code smells, we run linters, profilers, formatters, bounds checkers, documentation generators (for [libraries](https://ctc-oss.github.io/omega-edit/) at least), and use BDD (whenever we can -- UI development has proven to be a challenge here).  But what's worked best in my career is beelining to "working first".  Set aside the academics, optimizations, code poetry, endless design meetings, and just get the thing up and working, and then keep it working (swapping out the aircraft wings while in flight).  At the end of the day, this is an application for users and they don't care if the code is Shakespearian or bubblegum and duct-tape, as long as it _works_.  That's not to say there isn't room for great design, there is, but that comes over time as the full application takes shape and we can prioritize the pain points and address them along the way.  Since UI development isn't something Robert and I have a ton of experience in (TypeScript, and Svelte we learned on this project) experience told us that we're not going to nail the design right off the bat, but we do have road map items and a rough schedule that we intend on sticking to.

**Review by rthomas320 — APPROVED** [review]
Pulled down the PR and ran in a windows environment.  The change works as described.
